In [2]:
from keras.datasets import mnist # type: ignore
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [11]:
print(f"The shape of train images are: {train_images.shape}")
print(f"The number of trainin labels are: {len(train_labels)}")
print("--" * 30)
print(f"The shape of test labels are: {len(test_labels)}")
test_labels

The shape of train images are: (60000, 28, 28)
The number of trainin labels are: 60000
------------------------------------------------------------
The shape of test labels are: 10000


array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

## Workflow
We feed the training data (train_images and train_labels) to the NN. At this stage, the NN will try to associate the images and the labels. Then, we ask the NN to make predictions for test images, and we need to verify those predictions using test_labels.

In [12]:
import keras
from keras import layers

model = keras.Sequential(
    [
        layers.Dense(512, activation="relu"),
        layers.Dense(10, activation="softmax")
    ]
)

#### Note:
Layer is the core building blocks of NN. It acts as a data filter. `Data --> Layer --> Data (useful form).`

The model above, consisted of the following:
- Two `Dense` layers, fully connected neural layers. 
- The second layer, is a 10-way softmax classification layer means that it will return an array of 10 probabilities scores (summing to 1). Each score representing the probability of current digit image, belongs to one of 10 digit classes.

**What is next?**

To make the model ready for training, we need to pick three more things, as part of compilation step:
1) _A loss function_, This is how the model will be able to measure its performance on the training data, and thus how it be able to steer itself on the right direction.
2) _An Opitmizer_, This is a mechanism of the model to update itself based on the training data it sees, this is done to improve its performance.
3) _Metrics to monitor during training and testing_, The accuracy, how many images that were correctly classified. 

In [13]:
model.compile(
    optimizer="adam",
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

### Preprocessing data
Before we train our model, we need to preprocess the data as the our model expected. We cannot feed the model with data that the model doed not expect. This is done by reshaping it into the shape that the model expect and scaling it so that all values are in the interval of [0, 1]. 
Previously, our training data were stored in an array of shape `[60000, 28, 28]` of type `uint8` with values in the `[0, 255]` interval. We will transform it into a `float32` array of shape (`60000, 28 8 28`) with values between 0 and 1.

In [15]:
train_images = train_images.reshape((60000, 28 * 28)) # reshaping
train_images = train_images.astype("float32") / 255 # scaling
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

The data is now ready for the model and we can just train our model with the preprocessed data. We can do it by using a call to the model's `fit()` method. 

In [16]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 4s 7ms/step - accuracy: 0.8686 - loss: 0.4664
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9672 - loss: 0.1152
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9783 - loss: 0.0746
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9865 - loss: 0.0465
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - accuracy: 0.9906 - loss: 0.0345


#### Use the trained model for prediction
Our model has been trained. The accuracy is 0.9906 (99.06%) on the training data which is wow. Since we have our trained model, we can now use the model to predict class probabilities for new digits. 

In [17]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
predictions[0]

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 40ms/step


array([4.0499768e-07, 3.6980744e-08, 4.7164144e-06, 2.0174327e-04,
       3.7871689e-10, 9.9578085e-07, 3.3231508e-11, 9.9978715e-01,
       6.9647979e-07, 4.2309766e-06], dtype=float32)

Each number of index i in that array corresponds to the probability that digit image
test_digits[0] belong to class i.
This first test digit has the highest probability score (0.99999106, almost 1) at index
7, so according to our model, it must be a 7.

In [20]:
predictions[0].argmax()

7

In [21]:
predictions[0][7]

0.99978715

Check if the labels agree?


In [22]:
test_labels[0]

7

In [18]:
predictions[1]

array([2.3571848e-08, 7.0978681e-06, 9.9997580e-01, 1.3372930e-05,
       1.6259679e-12, 3.4135070e-07, 4.4610438e-08, 7.2260963e-13,
       3.2665839e-06, 3.0537024e-11], dtype=float32)

In [25]:
print(f"Model prediction for digit at index 1: {predictions[1].argmax()}")
print(f"Model prediction probability for index 1: {predictions[1][2]}")
print(f"Test label for digit at index 1: {test_labels[1]}")

Model prediction for digit at index 1: 2
Model prediction probability for index 1: 0.9999758005142212
Test label for digit at index 1: 2
